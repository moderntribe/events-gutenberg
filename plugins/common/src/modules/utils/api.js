import { config } from '@moderntribe/common/utils/globals';

/**
 * Returns a URL that is part of the WP-JSON endpoint associated with the site where the
 * extensions is being used.
 *
 * @param {string} path The path pointing to the endpoint
 * @param {string} namespace The namespace location of the endpoint
 * @returns {string} The URL used to hit the endpoint
 */
export const endpointUrl = ( path = '', namespace = 'wp/v2' ) => {
	const tribe = config();
	const rest = tribe.rest || {};
	const { url = '' } = rest;
	const namespaces = rest.namespaces || {};
	const core = namespaces.core || namespace;
	return `${ url }${ core }${ path }`;
};
