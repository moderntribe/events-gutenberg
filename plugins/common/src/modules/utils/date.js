/**
 * External dependencies
 */
import { get, identity } from 'lodash';

/**
 * Wordpress dependencies
 */
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import {
	moment as momentUtil,
	timezone as timezoneUtil,
} from '@moderntribe/common/utils';

const WPDateSettings = get( window, 'tribe_date_settings', {} );
const { formats = {}, timezone = {} } = WPDateSettings;

export const FORMATS = {
	TIME: 'HH:mm:ss',
	DATE_TIME: 'YYYY-MM-DD HH:mm:ss',
	WP: {
		time: 'g:i a',
		time24Hr: 'H:i',
		date: 'F j, Y',
		datetime: 'F j, Y g:i a',
		dateNoYear: 'F j',
		...formats,
	},
	TIMEZONE: {
		string: 'UTC',
		...timezone,
	},
	DATABASE: {
		date: 'Y-m-d',
		datetime: 'Y-m-d H:i:s',
		time: 'H:i:s',
	},
};

export const TODAY = new Date();

export const timezonesAsSelectData = () => {
	return timezones().map( ( tzone ) => ( {
		value: tzone.key,
		label: tzone.text,
	} ) );
};

export const timezones = () => {
	return timezoneUtil.getItems()
		.map( ( group ) => group.options || [] )
		.reduce( ( prev, current ) => [ ...prev, ...current ], [] );
};

export const toNaturalLanguage = ( params = {} ) => {
	const options = {
		date: null,
		format: {
			month: 'MMM',
			day: 'D',
			year: 'YYYY',
			time: 'h:mm a',
		},
		separator: '',
		...params,
	};

	const parsed = {
		text: '',
		moment: options.date && momentUtil.toMoment( options.date ),
		detail: {
			day: '',
			month: '',
			year: '',
			time: '',
		},
		isValid: false,
	};

	parsed.isValid = Boolean( parsed.moment && parsed.moment.isValid() );

	if ( parsed.isValid ) {
		parsed.detail = {
			month: `${ parsed.moment.format( options.format.month ) }`,
			day: `${ parsed.moment.format( options.format.day ) }`,
			year: `${ parsed.moment.format( options.format.year ) }`,
			time: `${ parsed.moment.format( options.format.time ) }`,
		};
		const { detail } = parsed;
		parsed.text = `${ detail.month } ${ detail.day } ${ detail.year } ${ options.separator } ${ detail.time }`;
	}
	return parsed;
};

export const rangeToNaturalLanguage = ( start = '', end = '', separators = {} ) => {
	const separatorOptions = {
		time: __( 'at', 'events-gutenberg' ),
		date: ' - ',
		...separators,
	}
	const from = toNaturalLanguage( { date: start, separator: separatorOptions.time } );
	const to = toNaturalLanguage( { date: end, separator: separatorOptions.time } );
	const parts = [ from.text ];

	if ( from.isValid && to.isValid ) {
		if ( momentUtil.isSameDay( from.moment, to.moment ) ) {
			/**
			 * If both dates are happening on the same day only the time is relevant on the secnod
			 * part of the string to keep string cleaner like: 'Oct 8 2018 at 12:00 pm - 12:30 pm'
			 * instead of 'Oct 8 2018 at 12:00 pm - Oct 8 2018 at 12:30 pm'
			 */
			parts.push( to.detail.time );
		} else if ( momentUtil.isSameMonth( from.moment, to.moment ) ) {
			/**
			 * If both dates are happening on the same month and not on the same day
			 * we don't need to show the same year twice like:  'Oct 8 2018 at 12:00 pm - Dec 20 12:30 pm'
			 * instead of  'Oct 8 2018 at 12:00 pm - December 24 2018 12:30 pm'
			 */
			parts.push( `${ to.detail.month } ${ to.detail.day } at ${ to.detail.time }` );
		} else {
			// Otherwise just use the full text
			parts.push( to.text );
		}
	}

	return parts.filter( identity ).join( separatorOptions.date );
};
