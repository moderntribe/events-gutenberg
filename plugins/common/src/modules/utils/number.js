/**
 * Calculate the percentage of two numbers
 *
 * @param {number} value Initial value from where to take the percentage
 * @param {number} total Total value to get the percentage relative to this value
 * @returns {number} total percentage value
 */
export const percentage = ( value = 0, total = 0 ) => {
	if ( total <= 0 ) {
		return 0;
	}

	const result = Math.floor( ( value / total ) * 100 );

	if ( isNaN( result ) ) {
		return 0;
	}

	// ensure number does not goes out of 100%
	return Math.min( 100, result );
};
