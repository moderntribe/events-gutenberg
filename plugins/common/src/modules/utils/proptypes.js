/**
 * Create chainable validator with isRequired check
 * @param {function} validator
 */
export const createChainableValidator = ( validator ) => {
	const createChainedValidator = (
		isRequired,
		props,
		propName,
		componentName,
	) => {
		const propValue = props[ propName ];

		if ( propValue == null ) {
			if ( isRequired ) {
				if ( propValue === null ) {
					/* eslint-disable-next-line max-len */
					return new Error( 'The prop `' + propName + '` is marked as required in `' + componentName + '`, but its value is `null`.' );
				}
				/* eslint-disable-next-line max-len */
				return new Error( 'The prop `' + propName + '` is marked as required in `' + componentName + '`, but its value is `undefined`.' );
			}
			return null;
		} else {
			return validator( props, propName, componentName );
		}
	};

	const chainedValidator = createChainedValidator.bind( null, false );
	chainedValidator.isRequired = createChainedValidator.bind( null, true );

	return chainedValidator;
}

/**
 * PropTypes check for type string and time format
 * @param {object} props
 * @param {string} propName
 * @param {string} componentName
 */
export const timeFormat = ( props, propName, componentName ) => {
	const propValue = props[ propName ];

	if ( typeof propValue !== 'string' ) {
		const type = typeof propValue;
		/* eslint-disable-next-line max-len */
		return new Error('Invalid prop `' + propName + '` of type `' + type + '` supplied to `' + componentName + '`, expected `string`.' );
	}

	const timeRegex = /^([01]?[0-9]|2[0-3]):[0-5][0-9]$/;
	if ( ! timeRegex.test( propValue ) ) {
		/* eslint-disable-next-line max-len */
		return new Error('Invalid prop `' + propName + '` shape supplied to `' + componentName + '`, expected `hh:mm`.' );
	}

	return null;
};
