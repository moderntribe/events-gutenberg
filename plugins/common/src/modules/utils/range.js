/**
 * External dependencies
 */
import { trim, isEmpty } from 'lodash';

/**
 * Parse a string into a range type of string {a} - {b} where a and b are numbers
 *
 * @param {string} input The original string
 * @returns {string} A formatted range string.
 */
export const parser = ( input ) => {
	const range = trim( input );

	if ( isEmpty( range ) ) {
		return range;
	}

	const chars = parseChars( input );

	if ( isEmpty( chars ) ) {
		return chars;
	}

	const [ a, b ] = extractParts( chars );
	const [ num_a, num_b ] = [ parseFloat( a ), parseFloat( b ) ];

	if ( ! num_b || num_b === num_a ) {
		return num_a === 0 ? '' : trim( a );
	}

	return num_a >= num_b ? `${ trim( b ) } - ${ trim( a ) }` : `${ trim( a ) } - ${ trim( b ) }`;
};

/**
 * Remove any char that is not a: number, dash, dot or comma.
 *
 * @param {string} input The string from where to extract the chars
 * @returns {string} A string with only valid chars
 */
export const parseChars = ( input ) => (
	input
		.split( ' ' )
		.map( ( part ) => {
			// Remove anything that is not a number a period or a dash
			return part.replace( /[^0-9.,-]/g, '' );
		} )
		.join( ' ' )
		.trim()
);

/**
 * Extract only valid numbers from the string
 *
 * @param {string} chars The chars to be split into parts.
 * @returns {array} An array with the parts
 */
export const extractParts = ( chars ) => (
	chars
	// Convert , into . so we can parse into numbers
		.replace( /,/g, '.' )
		.split( '-' )
		.map( ( item ) => {
			const re = /([0-9]+(.[0-9]+)?)/g;
			const result = re.exec( item.trim() );
			return null === result ? '' : result[ 1 ];
		} )
		.filter( ( item ) => ! isEmpty( item ) )
		.map( ( item ) => {
			// If the user input the price with decimals (even .00) we want to keep them
			const decimals = 0 < item.indexOf( '.' ) ? 2 : 0 ;
			return parseFloat( item ).toFixed( decimals );
		} )
		.filter( ( item ) => ! isNaN( item ) )
		.slice( 0, 2 )
);

/**
 * Test to see if an input range is free of cost
 *
 * @param {string} input Range input
 * @returns {boolean} true if the event has 0 on all parts of the range, false otherwise
 */
export const isFree = ( input ) => {
	const parts = input.split( '-' );

	const test = parts
		.map( ( item ) => parseFloat( item ) )
		.filter( ( item ) => ! isNaN( item ) )
		.filter( ( item ) => item === 0 );

	return parts.length === test.length;
};
