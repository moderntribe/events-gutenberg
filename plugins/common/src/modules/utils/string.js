/**
 * External dependencies
 */
import { escapeRegExp, isUndefined } from 'lodash';

/**
 * Test if a string is equivalent to a true value
 *
 * @param {string} value The value to be tested
 * @returns {boolean} true if the value is a valid "true" value.
 */
export const isTruthy = ( value ) => {
	const validValues = [
		'true',
		'yes',
		'1',
	];
	return validValues.indexOf( value ) !== - 1;
};

/**
 * Test if a string is equivalent to a false value
 *
 * @param {string} value The value to be tested
 * @returns {boolean} true if the value is a valid "false" value.
 */
export const isFalsy = ( value ) => {
	const validValues = [
		'false',
		'no',
		'0',
		'',
	];
	return validValues.indexOf( value ) !== - 1;
};

export const replaceWithObject = ( str = '', pairs = {} ) => {
	const substrs = Object.keys( pairs ).map( escapeRegExp );
	return str.split( RegExp( `(${ substrs.join( '|' ) })` ) )
		.map( part => isUndefined( pairs[ part ] ) ? part : pairs[ part ] )
		.join( '' );
};

/**
 * Interpolate any number of %d on number values
 *
 * @param {string} str The input with the placeholders
 * @param {number} params Any list of numbers to replace with placeholders
 * @returns {string} The updated string
 */
export const interpolateNumbers = ( str = '', ...params ) => {
	const [ number, ...remaining ] = params;

	const position = str.indexOf( '%d' );
	if ( position === -1 || params.length === 0 ) {
		return str;
	}

	const chars = str.split( '' );
	if ( chars[ position ] === '%' && chars[ position + 1 ] === 'd' ) {
		// Remove %d at start of % + 2 steps until d and insert number on it's place.
		chars.splice( position, 2, number );
		return interpolateNumbers( chars.join( '' ), ...remaining );
	} else {
		return chars.join( '' );
	}
};
