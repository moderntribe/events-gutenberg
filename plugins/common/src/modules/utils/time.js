/**
 * External dependencies
 */
import zeroFill from 'zero-fill';

export const MINUTE_IN_SECONDS = 60;
export const HALF_HOUR_IN_SECONDS = MINUTE_IN_SECONDS * 30;
export const HOUR_IN_SECONDS = 60 * MINUTE_IN_SECONDS;
export const DAY_IN_SECONDS = 24 * HOUR_IN_SECONDS;

export const START_OF_DAY = '00:00';
export const END_OF_DAY = '23:59';

/**
 * The code below is copied from the library hh-mm-ss
 * Link: https://www.npmjs.com/package/hh-mm-ss
 * The code has been copied so that Modern Tribe can maintain this library
 * internally and adjust as needed.
 */
export const TIME_FORMAT_HH_MM_SS_SSS = 'hh:mm:ss.sss';
export const TIME_FORMAT_HH_MM_SS = 'hh:mm:ss';
export const TIME_FORMAT_HH_MM = 'hh:mm';
export const TIME_FORMAT_MM_SS_SSS = 'mm:ss.sss';
export const TIME_FORMAT_MM_SS = 'mm:ss';

export const SECOND_IN_MS = 1000;
export const MINUTE_IN_MS = MINUTE_IN_SECONDS * SECOND_IN_MS;
export const HOUR_IN_MS = HOUR_IN_SECONDS * SECOND_IN_MS;

/**
 * Converts milliseconds to time in the format provided
 *
 * @param {int} ms Milliseconds to convert from
 * @param {string} format Format of time to convert to
 * @returns {string} Time string equivalent of milliseconds in format provided
 */
export const fromMilliseconds = ( ms, format = TIME_FORMAT_MM_SS ) => {
	if ( typeof ms !== 'number' || Number.isNaN( ms ) ) {
		/* eslint-disable-next-line max-len */
		throw new Error( 'Argument `ms` provided to `fromMilliseconds` is not a number or is NaN.' );
	}

	let absMs = Math.abs( ms );

	let negative = ( ms < 0 );
	let hours = Math.floor( absMs / HOUR_IN_MS );
	let minutes = Math.floor( absMs % HOUR_IN_MS / MINUTE_IN_MS );
	let seconds = Math.floor( absMs % MINUTE_IN_MS / SECOND_IN_MS );
	let miliseconds = Math.floor( absMs % SECOND_IN_MS );

	return formatTime( {
		negative,
		hours,
		minutes,
		seconds,
		miliseconds,
	}, format );
};

/**
 * Converts seconds to time in the format provided
 *
 * @param {int} s Seconds to convert from
 * @param {string} format Format of time to convert to
 * @returns {string} Time string equivalent of seconds in format provided
 */
export const fromSeconds = ( s, format = TIME_FORMAT_MM_SS ) => {
	if ( typeof s !== 'number' || Number.isNaN( s ) ) {
 		/* eslint-disable-next-line max-len */
		throw new Error( 'Argument `s` provided to `fromSeconds` is not a number or is NaN.' );
	}

	let ms = s * SECOND_IN_MS;

	return fromMilliseconds( ms, format );
};

/**
 * Converts time in the format provided to milliseconds
 *
 * @param {string} time Time string to convert from
 * @param {string} format Format of time to convert from
 * @returns {int} Milliseconds equivalent of time string in format provided
 */
export const toMilliseconds = ( time, format = TIME_FORMAT_MM_SS ) => {
	let re;

	if ( [
		TIME_FORMAT_HH_MM_SS_SSS,
		TIME_FORMAT_HH_MM_SS,
		TIME_FORMAT_MM_SS_SSS,
		TIME_FORMAT_MM_SS,
	].includes( format ) ) {
		re = /^(-)?(?:(\d\d+):)?(\d\d):(\d\d)(\.\d+)?$/;
	} else if ( format === TIME_FORMAT_HH_MM ) {
		re = /^(-)?(\d\d):(\d\d)(?::(\d\d)(?:(\.\d+))?)?$/;
	} else {
 		/* eslint-disable-next-line max-len */
		throw new Error( 'Argument `format` provided to `toMilliseconds` is not a recognized format.' );
	}

	let result = re.exec( time );
	if ( ! result ) {
		/* eslint-disable-next-line max-len */
		throw new Error( 'Argument `time` provided to `toMilliseconds` is not a recognized format.' );
	}

	let negative = result[ 1 ] === '-';
	let hours = result[ 2 ] | 0;
	let minutes = result[ 3 ] | 0;
	let seconds = result[ 4 ] | 0;
	let miliseconds = Math.floor( 1000 * result[ 5 ] | 0 );

	if ( minutes >= 60 || seconds >= 60 ) {
		/* eslint-disable-next-line max-len */
		throw new Error( 'Argument `time` provided to `toMilliseconds` contains minutes or seconds greater than 59.' );
	}

	return ( negative ? -1 : 1 ) * (
		hours * HOUR_IN_MS
		+ minutes * MINUTE_IN_MS
		+ seconds * SECOND_IN_MS
		+ miliseconds
	);
};

/**
 * Converts time in the format provided to seconds
 *
 * @param {string} time Time string to convert from
 * @param {string} format Format of time to convert from
 * @returns {int} Seconds equivalent of time string in format provided
 */
export const toSeconds = ( time, format = TIME_FORMAT_MM_SS ) => {
	let ms = toMilliseconds( time, format );
	return Math.floor( ms / SECOND_IN_MS );
};

/**
 * Formats time object to time string in the format provided
 *
 * @param {object} time Time object to format from
 * @param {string} format Format of time to format to
 * @returns {string} Time string in format provided
 */
export const formatTime = ( time, format ) => {
	let showMs;
	let showSc;
	let showHr;

	switch ( format ) {
		case TIME_FORMAT_HH_MM_SS_SSS:
			showMs = true;
			showSc = true;
			showHr = true;
			break;
		case TIME_FORMAT_HH_MM_SS:
			showMs = ! ! time.miliseconds;
			showSc = true;
			showHr = true;
			break;
		case TIME_FORMAT_HH_MM:
			showMs = ! ! time.miliseconds;
			showSc = showMs || ! ! time.seconds;
			showHr = true;
			break;
		case TIME_FORMAT_MM_SS_SSS:
			showMs = true;
			showSc = true;
			showHr = ! ! time.hours;
			break;
		case TIME_FORMAT_MM_SS:
			showMs = ! ! time.miliseconds;
			showSc = true;
			showHr = ! ! time.hours;
			break;
		default:
			/* eslint-disable-next-line max-len */
			throw new Error( 'Argument `format` provided to `formatTime` is not a recognized format.' );
	}

	let hh = zeroFill( 2, time.hours );
	let mm = zeroFill( 2, time.minutes );
	let ss = zeroFill( 2, time.seconds );
	let sss = zeroFill( 3, time.miliseconds );

	return ( time.negative ? '-' : '' ) + ( showHr ? (
		showMs ? `${hh}:${mm}:${ss}.${sss}` : showSc ? `${hh}:${mm}:${ss}` : `${hh}:${mm}`
	) : (
		showMs ? `${mm}:${ss}.${sss}` : `${mm}:${ss}`
	) );
};
