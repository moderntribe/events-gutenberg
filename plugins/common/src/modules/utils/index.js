import * as date from './date';
import * as dom from './dom';
import * as globals from './globals';
import * as moment from './moment';
import * as time from './time';
import * as timezone from './timezone';

export { date };
export { dom };
export { globals };
export { moment };
export { time };
export { timezone };
export { default as TribePropTypes } from './proptypes';
