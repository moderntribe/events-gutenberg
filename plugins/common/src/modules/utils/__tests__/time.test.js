/**
 * Internal dependencies
 */
import { time } from '@moderntribe/common/utils';

describe( 'Tests for time.js', () => {
	test( 'MINUTE_IN_SECONDS', () => {
		expect( time.MINUTE_IN_SECONDS ).toEqual( 60 );
	} );

	test( 'HALF_HOUR_IN_SECONDS', () => {
		expect( time.HALF_HOUR_IN_SECONDS ).toEqual( 1800 );
	} );

	test( 'HOUR_IN_SECONDS', () => {
		expect( time.HOUR_IN_SECONDS ).toEqual( 3600 );
	} );

	test( 'DAY_IN_SECONDS', () => {
		expect( time.DAY_IN_SECONDS ).toEqual( 86400 );
	} );

	test( 'START_OF_DAY', () => {
		expect( time.START_OF_DAY ).toEqual( '00:00' );
	} );

	test( 'END_OF_DAY', () => {
		expect( time.END_OF_DAY ).toEqual( '23:59' );
	} );

	test( 'TIME_FORMAT_HH_MM_SS_SSS', () => {
		expect( time.TIME_FORMAT_HH_MM_SS_SSS ).toEqual( 'hh:mm:ss.sss' );
	} );

	test( 'TIME_FORMAT_HH_MM_SS', () => {
		expect( time.TIME_FORMAT_HH_MM_SS ).toEqual( 'hh:mm:ss' );
	} );

	test( 'TIME_FORMAT_HH_MM', () => {
		expect( time.TIME_FORMAT_HH_MM ).toEqual( 'hh:mm' );
	} );

	test( 'TIME_FORMAT_MM_SS_SSS', () => {
		expect( time.TIME_FORMAT_MM_SS_SSS ).toEqual( 'mm:ss.sss' );
	} );

	test( 'TIME_FORMAT_MM_SS', () => {
		expect( time.TIME_FORMAT_MM_SS ).toEqual( 'mm:ss' );
	} );

	test( 'HOUR_IN_MS', () => {
		expect( time.HOUR_IN_MS ).toEqual( 3600000 );
	} );

	test( 'MINUTE_IN_MS', () => {
		expect( time.MINUTE_IN_MS ).toEqual( 60000 );
	} );

	test( 'SECOND_IN_MS', () => {
		expect( time.SECOND_IN_MS ).toEqual( 1000 );
	} );

	/**
	 * Below are tests copied from the hh-mm-ss library and adjusted to use
	 * Jest instead of Tape for testing.
	 * Link: https://github.com/Goldob/hh-mm-ss/blob/master/test/index.js
	 */
	test( 'fromMs() test', () => {
		// Basic functionality
		expect( time.fromMs( 75000 ) ).toEqual( '01:15' );
		expect( time.fromMs( 442800000 ) ).toEqual( '123:00:00' );
		expect( time.fromMs( 90576 ) ).toEqual( '01:30.576' );
		expect( time.fromMs( -157250 ) ).toEqual( '-02:37.250' );

		// Output formatting
		expect( time.fromMs( 38000, 'mm:ss.sss' ) ).toEqual( '00:38.000' );
		expect( time.fromMs( 0, 'hh:mm:ss' ) ).toEqual( '00:00:00' );
		expect( time.fromMs( 3600000, 'mm:ss' ) ).toEqual( '01:00:00' );
		expect( time.fromMs( 4500000, 'hh:mm' ) ).toEqual( '01:15' );
		expect( time.fromMs( -9900000, 'hh:mm' ) ).toEqual( '-02:45' );

		// Input validation
		expect( () => time.fromMs( null ) ).toThrowErrorMatchingSnapshot();
		expect( () => time.fromMs('text') ).toThrowErrorMatchingSnapshot();
		expect( () => time.fromMs(0, 'mm:hh:ss') ).toThrowErrorMatchingSnapshot();
	} );

	test( 'fromS() test', () => {
		// Basic functionality
		expect( time.fromS( 75 ) ).toEqual( '01:15' );
		expect( time.fromS( 442800 ) ).toEqual( '123:00:00' );
		expect( time.fromS( -442800 ) ).toEqual( '-123:00:00' );

		// Output formatting
		expect( time.fromS( 38, 'mm:ss.sss' ) ).toEqual( '00:38.000' );
		expect( time.fromS( 0, 'hh:mm:ss' ) ).toEqual( '00:00:00' );
		expect( time.fromS( 3600, 'mm:ss' ) ).toEqual( '01:00:00' );
		expect( time.fromS( 4500, 'hh:mm' ) ).toEqual( '01:15' );
		expect( time.fromS( -9900, 'hh:mm' ) ).toEqual( '-02:45' );

		// Input validation
		expect( () => time.fromS( null ) ).toThrowErrorMatchingSnapshot();
		expect( () => time.fromS( 'text' ) ).toThrowErrorMatchingSnapshot();
		expect( () => time.fromS( 0, 'mm:hh:ss' ) ).toThrowErrorMatchingSnapshot();
	} );

	test( 'toMs() test', () => {
		// Basic functionality
		expect( time.toMs( '01:05:17' ) ).toEqual( 3917000 );
		expect( time.toMs( '137:00:00.0' ) ).toEqual( 493200000 );
		expect( time.toMs( '00:10.230' ) ).toEqual( 10230 );
		expect( time.toMs( '00:00:07.10845' ) ).toEqual( 7108 );
		expect( time.toMs( '-02:07:12' ) ).toEqual( -7632000 );
		expect( time.toMs( '02:00' ) ).toEqual( 120000 );
		expect( time.toMs( '02:00', 'hh:mm' ) ).toEqual( 7200000 );
		expect( time.toMs( '-04:35', 'hh:mm' ) ).toEqual( -16500000 );
		expect( time.toMs( '00:00:07.10845', 'hh:mm' ) ).toEqual( 7108 );

		// Input validation
		expect( () => time.toMs( '13:05:02:11' ) ).toThrowErrorMatchingSnapshot();
		expect( () => time.toMs( '153' ) ).toThrowErrorMatchingSnapshot();
		expect( () => time.toMs( null ) ).toThrowErrorMatchingSnapshot();
		expect( () => time.toMs( '00:62' ) ).toThrowErrorMatchingSnapshot();
		expect( () => time.toMs( '01:30', 'mm:hh:ss' ) ).toThrowErrorMatchingSnapshot();
	} );

	test( 'toS() test', () => {
		// Basic functionality
		expect( time.toS( '01:05:17' ) ).toEqual( 3917 );
		expect( time.toS( '137:00:00.0' ) ).toEqual( 493200 );
		expect( time.toS( '00:10.230' ) ).toEqual( 10 );
		expect( time.toS( '00:00:07.10845' ) ).toEqual( 7 );
		expect( time.toS( '-02:07:12' ) ).toEqual( -7632 );
		expect( time.toS( '02:00' ) ).toEqual( 120 );
		expect( time.toS( '02:00', 'hh:mm' ) ).toEqual( 7200 );
		expect( time.toS( '-04:35', 'hh:mm' ) ).toEqual( -16500 );
		expect( time.toS( '00:00:07.10845', 'hh:mm' ) ).toEqual( 7 );

		// Input validation
		expect( () => time.toS( '13:05:02:11' ) ).toThrowErrorMatchingSnapshot();
		expect( () => time.toS( '153' ) ).toThrowErrorMatchingSnapshot();
		expect( () => time.toS( null ) ).toThrowErrorMatchingSnapshot();
		expect( () => time.toS( '00:62' ) ).toThrowErrorMatchingSnapshot();
		expect( () => time.toS( '01:30', 'mm:hh:ss' ) ).toThrowErrorMatchingSnapshot();
	} );

	test( 'symmetrical conversion test', () => {
		/*
		 * fromMs() and toMs() for all formats
		 */

		// 90000ms = 1m 30s
		expect( time.toMs( time.fromMs( 90000, 'mm:ss' ), 'mm:ss' ) ).toEqual( 90000 );
		expect( time.toMs( time.fromMs( 90000, 'mm:ss.sss' ), 'mm:ss.sss' ) ).toEqual( 90000 );
		expect( time.toMs( time.fromMs( 90000, 'hh:mm' ), 'hh:mm' ) ).toEqual( 90000 );
		expect( time.toMs( time.fromMs( 90000, 'hh:mm:ss' ), 'hh:mm:ss' ) ).toEqual( 90000 );
		expect( time.toMs( time.fromMs( 90000, 'hh:mm:ss.sss' ), 'hh:mm:ss.sss' ) ).toEqual( 90000 );

		// 7517245ms = 2h 5m 17.245s
		expect( time.toMs( time.fromMs( 7517245, 'mm:ss' ), 'mm:ss' ) ).toEqual( 7517245 );
		expect( time.toMs( time.fromMs( 7517245, 'mm:ss.sss' ), 'mm:ss.sss' ) ).toEqual( 7517245 );
		expect( time.toMs( time.fromMs( 7517245, 'hh:mm' ), 'hh:mm' ) ).toEqual( 7517245 );
		expect( time.toMs( time.fromMs( 7517245, 'hh:mm:ss' ), 'hh:mm:ss' ) ).toEqual( 7517245 );
		expect( time.toMs( time.fromMs( 7517245, 'hh:mm:ss.sss' ), 'hh:mm:ss.sss' ) ).toEqual( 7517245 );

		// -10800000ms = -3h
		expect( time.toMs( time.fromMs( -10800000, 'mm:ss' ), 'mm:ss' ) ).toEqual( -10800000 );
		expect( time.toMs( time.fromMs( -10800000, 'mm:ss.sss' ), 'mm:ss.sss' ) ).toEqual( -10800000 );
		expect( time.toMs( time.fromMs( -10800000, 'hh:mm' ), 'hh:mm' ) ).toEqual( -10800000 );
		expect( time.toMs( time.fromMs( -10800000, 'hh:mm:ss' ), 'hh:mm:ss' ) ).toEqual( -10800000 );
		expect( time.toMs( time.fromMs( -10800000, 'hh:mm:ss.sss' ), 'hh:mm:ss.sss' ) ).toEqual( -10800000 );

		/*
		 * fromS() and toMs() for all formats
		 */

		// 930s = 15m 30s
		expect( time.toS( time.fromS( 930, 'mm:ss' ), 'mm:ss') ).toEqual( 930 );
		expect( time.toS( time.fromS( 930, 'mm:ss.sss' ), 'mm:ss.sss') ).toEqual( 930 );
		expect( time.toS( time.fromS( 930, 'hh:mm' ), 'hh:mm') ).toEqual( 930 );
		expect( time.toS( time.fromS( 930, 'hh:mm:ss' ), 'hh:mm:ss') ).toEqual( 930 );
		expect( time.toS( time.fromS( 930, 'hh:mm:ss.sss' ), 'hh:mm:ss.sss') ).toEqual( 930 );

		// 4850s = 1h 20m 50s
		expect( time.toS( time.fromS( 4850, 'mm:ss' ), 'mm:ss') ).toEqual( 4850 );
		expect( time.toS( time.fromS( 4850, 'mm:ss.sss' ), 'mm:ss.sss') ).toEqual( 4850 );
		expect( time.toS( time.fromS( 4850, 'hh:mm' ), 'hh:mm') ).toEqual( 4850 );
		expect( time.toS( time.fromS( 4850, 'hh:mm:ss' ), 'hh:mm:ss') ).toEqual( 4850 );
		expect( time.toS( time.fromS( 4850, 'hh:mm:ss.sss' ), 'hh:mm:ss.sss') ).toEqual( 4850 );

		// -300s = -5m
		expect( time.toS( time.fromS( -300, 'mm:ss' ), 'mm:ss') ).toEqual( -300 );
		expect( time.toS( time.fromS( -300, 'mm:ss.sss' ), 'mm:ss.sss') ).toEqual( -300 );
		expect( time.toS( time.fromS( -300, 'hh:mm' ), 'hh:mm') ).toEqual( -300 );
		expect( time.toS( time.fromS( -300, 'hh:mm:ss' ), 'hh:mm:ss') ).toEqual( -300 );
		expect( time.toS( time.fromS( -300, 'hh:mm:ss.sss' ), 'hh:mm:ss.sss') ).toEqual( -300 );
	} );
} );
