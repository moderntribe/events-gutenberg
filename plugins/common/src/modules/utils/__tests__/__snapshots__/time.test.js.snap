// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for time.js fromMs() test 1`] = `"Argument \`ms\` provided to \`fromMs\` is not a number or is NaN."`;

exports[`Tests for time.js fromMs() test 2`] = `"Argument \`ms\` provided to \`fromMs\` is not a number or is NaN."`;

exports[`Tests for time.js fromMs() test 3`] = `"Argument \`format\` provided to \`toMs\` is not a recognized format."`;

exports[`Tests for time.js fromS() test 1`] = `"Argument \`s\` provided to \`fromS\` is not a number or is NaN."`;

exports[`Tests for time.js fromS() test 2`] = `"Argument \`s\` provided to \`fromS\` is not a number or is NaN."`;

exports[`Tests for time.js fromS() test 3`] = `"Argument \`format\` provided to \`toMs\` is not a recognized format."`;

exports[`Tests for time.js toMs() test 1`] = `"Argument \`time\` provided to \`toMs\` is not a recognized format."`;

exports[`Tests for time.js toMs() test 2`] = `"Argument \`time\` provided to \`toMs\` is not a recognized format."`;

exports[`Tests for time.js toMs() test 3`] = `"Argument \`time\` provided to \`toMs\` is not a recognized format."`;

exports[`Tests for time.js toMs() test 4`] = `"Argument \`time\` provided to \`toMs\` contains minutes or seconds greater than 60."`;

exports[`Tests for time.js toMs() test 5`] = `"Argument \`format\` provided to \`toMs\` is not a recognized format."`;

exports[`Tests for time.js toS() test 1`] = `"Argument \`time\` provided to \`toMs\` is not a recognized format."`;

exports[`Tests for time.js toS() test 2`] = `"Argument \`time\` provided to \`toMs\` is not a recognized format."`;

exports[`Tests for time.js toS() test 3`] = `"Argument \`time\` provided to \`toMs\` is not a recognized format."`;

exports[`Tests for time.js toS() test 4`] = `"Argument \`time\` provided to \`toMs\` contains minutes or seconds greater than 60."`;

exports[`Tests for time.js toS() test 5`] = `"Argument \`format\` provided to \`toMs\` is not a recognized format."`;
